docs/Developers/NextGen_Core.rst

```diff
@@ -0,0 +1,286 @@
+NextGen-Core
+============
+
+This document provides a brief overview of coala's NextGen-Core.
+coala's NextGen-Core comes with the promise of resolving the dependencies
+amongst bears more efficiently (by removing the queuing mechanism) and better
+performance.
+
+What is new?
+------------
+
+The following new features have been added as a part of the NextGen-Core:
+
+- Easier Interface
+- Official support for virtual files
+- Improved dependency system
+- New base bear type: ``DependencyBear``
+- Ability to modify bear dependencies at runtime
+- Custom parallelization for bears
+- Superior caching
+
+What has changed?
+-----------------
+
+- ``Global Bears`` are now called ``Project Bears`` and ``Local Bears`` are
+  now known as ``File Bears``. Both the ``ProjectBear`` and the ``FileBear``
+  classes inherit from the base class ``coalib.core.Bear``.
+- There is no need to pass ``HiddenResults`` between different bears (made
+  possible by the new dependency management system) to hide results to the
+  result callback. Only the results that were explicitly requested by passing
+  the needed bears are passed now. There is no need to pass instances of
+  ``Result`` anymore, any python object will do.
+- The former ``run`` function that was inherited by all the bears to run code
+  analysis is now replaced by the ``analyze`` function.
+
+Easier Interface
+----------------
+
+Running a coala session with the NextGen-Core is now as easy as
+``core.run(bears)``. The ``run`` method takes the following arguments ``bears``,
+``result_callback``, ``cache`` and ``executor`` (the last two are ``None`` by
+default) and initiates a coala session.
+
+* The ``bears`` argument contains the list of bears to be run for a coala
+  session.
+* The ``result_callback`` is a function that is called when the results from a
+  bear run are available.
+* The ``cache`` argument if provided enables caching and runs the session using
+  the cache provided to store the bear results.
+* The ``executor`` argument is used to provide a custom executor in which the
+  passed bears are to be run which is closed after the core is closed. If this
+  argument is not provided then ``ProcessPoolExecutor`` is used using as many
+  processes as cores available on the system.
+
+Official support for virtual files
+----------------------------------
+
+IDEs like IntelliJ use a virtual files to represent files in a filesystem (VFS)
+and perform operations on them. Hence NextGen-Core provides official support
+for virtual files, the bears have to point to the right file data objects when
+run, whether they are real files or virtual ones. This makes coala easier to
+integrate with IDEs since the contents reside in the IDE memory which is faster
+instead of reloading files from the disk.
```

Comment:
easier to integrate != faster :D

The reasoning of the part behind `since` doesn't match with the following content ;)

Reply:
Even I didn't understand this part fully I added it because it was there in review comment, I think I'll just leave this part out to avoid further confusion.